### PR TITLE
[REVIEW] Fix Warp Reduce calls in cuio statistics calculation to account for NaNs

### LINKED ISSUE #6365: [BUG] Incorrect float sum statistics calculation in ORC and Parquet
Prior to PR #6293 calculating float statistics used a [custom shuffle based warp reduce](https://github.com/kaatish/cudf/blob/57ef76927373d7260b6a0eda781e59a4c563d36e/cpp/src/io/statistics/column_stats.cu#L119) which avoided NaNs which is the expected behavior for stats calculation.

In the process of replacing those calls with `cub::WarpReduce` functions, checks for NaNs were overlooked for eg [here](https://github.com/kaatish/cudf/blob/b7b968899c939796e8bd2fc878a7e265ce93ee89/cpp/src/io/statistics/column_stats.cu#L279) and [here](https://github.com/kaatish/cudf/blob/b7b968899c939796e8bd2fc878a7e265ce93ee89/cpp/src/io/statistics/column_stats.cu#L577). This check needs to be added back to fix float sum statistics calculations.

### PR DESCRIPTION
**Pull request by kaatish**
Fixes #6365

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/src/io/statistics/column_stats.cu [hunk 1]
```diff
@@ -276,7 +276,9 @@ gatherFloatColumnStats(stats_state_s *s, statistics_dtype dtype, uint32_t t, Sto
       warp_reduce(storage.float_stats[t / 32]).Reduce(s->warp_max[t & 0x1f].fp_val, cub::Max());
     if (!(t & 0x1f)) { s->ck.max_value.fp_val = (vmax != 0.0) ? vmax : CUDART_ZERO; }
   } else if (t < 32 * 3) {
-    vsum = warp_reduce(storage.float_stats[t / 32]).Sum(s->warp_sum[t & 0x1f].fp_val);
+    double val = s->warp_sum[t & 0x1f].fp_val;
+    if (isnan(val)) { val = 0; }
+    vsum = warp_reduce(storage.float_stats[t / 32]).Sum(val);
```
Comment (vuule) [inline cpp/src/io/statistics/column_stats.cu:281]: This makes it consistent with the change above and removes the additional state.
```suggestion
    auto const& val = s->warp_sum[t & 0x1f].fp_val;
    vsum = warp_reduce(storage.float_stats[t / 32]).Sum(isnan(val) ? 0 : val);
```

File: cpp/src/io/statistics/column_stats.cu [hunk 2]
```diff
@@ -574,7 +576,9 @@ void __device__ mergeFloatColumnStats(merge_state_s *s,
       cub::WarpReduce<double>(storage.f64[t / 32]).Reduce(s->warp_max[t & 0x1f].fp_val, cub::Max());
     if (!(t & 0x1f)) { s->ck.max_value.fp_val = (vmax != 0.0) ? vmax : CUDART_ZERO; }
   } else if (t < 32 * 3) {
-    vsum = cub::WarpReduce<double>(storage.f64[t / 32]).Sum(s->warp_sum[t & 0x1f].fp_val);
+    double val = s->warp_sum[t & 0x1f].fp_val;
+    if (isnan(val)) { val = 0; }
+    vsum = cub::WarpReduce<double>(storage.f64[t / 32]).Sum(val);
```
Comment (vuule) [inline cpp/src/io/statistics/column_stats.cu:581]: Same here.

**Review by vuule — CHANGES_REQUESTED** [review]
Minor improvements suggested.
This aside, I still feel like using a custom Sum functor would be a better solution. We implement the logic once and call it in multiple places, rather than apply the same logic manually to all call sites. Functor seems useful in general (for example, if we use CUB block reductions here in the future).

**Comment by kaatish** [comment]
> Minor improvements suggested.
> This aside, I still feel like using a custom Sum functor would be a better solution. We implement the logic once and call it in multiple places, rather than apply the same logic manually to all call sites. Functor seems useful in general (for example, if we use CUB block reductions here in the future).

I have added the custom addition functor. There will be increased number of checks for NaNs compared to the previous solution but I do not think it will have a huge impact on performance.

**Review by davidwendt — APPROVED** [review]

**Review by vuule — APPROVED** [review]

**Review by OlivierNV — APPROVED** [review]
LGTM

**Comment by vuule** [comment]
rerun tests

**Pull request merged**